demos/dev_mode_key_provisioning/src/aws_dev_mode_key_provisioning.c

```diff
@@ -1027,24 +1031,34 @@ CK_RV xProvisionDevice( CK_SESSION_HANDLE xSession,
             ( CK_INVALID_HANDLE == xProvisionedState.xPublicKey ) ||
             ( NULL == xProvisionedState.pucDerPublicKey ) )
         {
-            /* Generate a new keyset if either of the above objects couldn't be
-             * found. */
-            xResult = xProvisionGenerateKeyPairEC( xSession,
-                                                   ( uint8_t * ) pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS,
-                                                   ( uint8_t * ) pkcs11configLABEL_DEVICE_PUBLIC_KEY_FOR_TLS,
-                                                   &xProvisionedState.xPrivateKey,
-                                                   &xProvisionedState.xPublicKey );
-
-            /* Get the bytes of the new public key. */
-            if( CKR_OK == xResult )
-            {
-                xGeneratedPrivateKey = CK_TRUE;
+            xKeysetGenerationMode = CK_TRUE;
+        }
 
-                prvExportPublicKey( xSession,
-                                    xProvisionedState.xPublicKey,
-                                    &xProvisionedState.pucDerPublicKey,
-                                    &xProvisionedState.ulDerPublicKeyLength );
-            }
+        /* Ignore errors. If any of the above objects couldn't be read, try to
+         * generate new ones below. */
+        xResult = CKR_OK;
+    }
+
+    #if ( 1 == keyprovisioningFORCE_GENERATE_NEW_KEYSET )
+        xKeysetGenerationMode = CK_TRUE;
+    #endif
+
+    if( ( xResult == CKR_OK ) && ( CK_TRUE == xKeysetGenerationMode ) )
+    {
+        /* Generate a new default keyset. */
+        xResult = xProvisionGenerateKeyPairEC( xSession,
+                                               ( uint8_t * ) pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS,
+                                               ( uint8_t * ) pkcs11configLABEL_DEVICE_PUBLIC_KEY_FOR_TLS,
+                                               &xProvisionedState.xPrivateKey,
+                                               &xProvisionedState.xPublicKey );
+
+        /* Get the bytes of the new public key. */
+        if( CKR_OK == xResult )
+        {
+            prvExportPublicKey( xSession,
```

Comment:
If we have a keypair already generated, and then we also use the FORCE GENERATE option, is there any chance we double allocate the public key bytes here (first in prvGetProvisionedState above, then again here)?

Reply:
Fixed, thank you.